### PR TITLE
P20-1183: Fix OmniAuth for Global Edition

### PR DESCRIPTION
**Pull request by artem-vavilov**
1. To make an OAuth callback accessible, it must be added to the whitelist of each SSO provider. Rather than repeating this process for every new Global Edition region, it is more efficient to strip the Global Edition prefix and handle the request as a standard route.
2. Prevents OmniAuth routes from being redirected to avoid disrupting the authentication process.

## Links
- JIRA task: [P20-1183](https://codedotorg.atlassian.net/browse/P20-1183)
- Original PR: https://github.com/code-dot-org/code-dot-org/pull/62061
- Related PR: https://github.com/code-dot-org/code-dot-org/pull/62858

### CODE REVIEW

File: lib/cdo/rack/global_edition.rb [hunk 1]
```diff
@@ -163,17 +162,18 @@ def call
       private def redirectable?(redirect_path)
         return false unless request.get? # only GET request can be redirected
         return false if request.xhr? # only non-AJAX requests should be redirected
+        return false if request.path.include?('/users/auth/')
```
Comment (artem-vavilov) [inline lib/cdo/rack/global_edition.rb:165]: Prevents OAuth routes from being redirected to the Global Edition's regional path (`/global/fa`)

File: dashboard/config/initializers/devise.rb [hunk 2]
```diff
@@ -354,6 +356,15 @@
 
   OmniAuth.config.before_request_phase do |env|
     request = Rack::Request.new(env)
+
+    # To make an OAuth URL valid, it should be added to the whitelist for each SSO provider.
+    # Instead of applying this for each new Global Edition region,
+    # it is more efficient to remove the Global Edition prefix and handle the path as a regular route.
+    if request.script_name && request.ge_region
```
Comment (artem-vavilov) [inline dashboard/config/initializers/devise.rb:363]: The changes only affect the Global Edition pages

File: dashboard/config/initializers/devise.rb [hunk 3]
```diff
@@ -354,6 +356,16 @@
...
   OmniAuth.config.before_request_phase do |env|
     request = Rack::Request.new(env)
+
+    # To make an OAuth callback accessible, it must be added to the whitelist of each SSO provider.
+    # Rather than repeating this process for every new Global Edition region,
+    # it is more efficient to strip the Global Edition prefix and handle the request as a standard route.
+    # See: +Rack::GlobalEdition+
+    if request.ge_region && request.script_name
+      ge_path_prefix = Cdo::GlobalEdition.path(request.ge_region)
+      request.script_name = request.script_name.sub(ge_path_prefix, '').presence
```
Comment (artem-vavilov) [inline dashboard/config/initializers/devise.rb:366]: Removes the Global Edition path prefix (e.g. `/global/fa`) from OAuth requests set in the middleware: https://github.com/code-dot-org/code-dot-org/blob/c6edd95da283907df4ff6c5c38343fcdfc756d4a/lib/cdo/rack/global_edition.rb#L55-L68

**Review by wilkie — APPROVED** [review]
Google sign up works on adhoc now for old signup flow.

Google sign up works on adhoc for new signup flow.

Microsoft sign up redirected properly but did not seem to continue the signup process. Not sure why... but it seemed to redirect perfectly fine which is an improvement. It might just be some weirdness where it will work in production.

I'm generally willing to assume the others work... they will be the same behavior as the non-regional sign up flow. I can't effectively test them, but we already know this won't break them.

Let's merge it and test the others on staging and/or prod and follow-up as needed.

Clever fix!

**Pull request merged**